src/Internal/Types/ScriptLookups.purs

```diff
@@ -1018,24 +1049,57 @@ processConstraint mpsMap osMap = do
           v <- liftM (CannotMakeValue cs tn i) (value i)
           _valueSpentBalancesOutputs <>= provideValue v
           pure $ map getNonAdaAsset $ value i
-      let
-        -- Create a redeemer with zero execution units then call Ogmios to
-        -- add the units in at the very end.
-        -- Hardcode minting policy index, then reindex with
-        -- `reindexMintRedeemers`.
-        redeemer = T.Redeemer
-          { tag: Mint
-          , index: zero
-          , data: unwrap red
-          , exUnits: zero
-          }
-      -- Remove mint redeemers from array before reindexing.
-      _redeemersTxIns %= filter \(T.Redeemer { tag } /\ _) -> tag /= Mint
-      -- Reindex mint redeemers.
-      mintRedeemers <- lift $ reindexMintRedeemers mpsHash redeemer
-      -- Append reindexed mint redeemers to array.
-      _redeemersTxIns <>= map (_ /\ Nothing) mintRedeemers
+
+      when (not isNative) $ do
+        let
+          -- Create a redeemer with zero execution units then call Ogmios to
+          -- add the units in at the very end.
+          -- Hardcode minting policy index, then reindex with
+          -- `reindexMintRedeemers`.
+          redeemer = T.Redeemer
+            { tag: Mint
+            , index: zero
+            , data: unwrap red
+            , exUnits: zero
+            }
+        -- Remove mint redeemers from array before reindexing.
+        _redeemersTxIns %= filter \(T.Redeemer { tag } /\ _) -> tag /= Mint
+        -- Reindex mint redeemers.
+        mintRedeemers <- lift $ reindexMintRedeemers mpsHash redeemer
+        -- Append reindexed mint redeemers to array.
+        _redeemersTxIns <>= map (_ /\ Nothing) mintRedeemers
+
+      _cpsToTxBody <<< _mint <>= map wrap mintVal
+
+    -- TODO: amir decide whether to split MustMintValue & MustMintValueUsingNativeScript
+    MustMintValueUsingNativeScript nsHash tn i -> runExceptT do
+      let mpHash = wrap <<< unwrap $ nsHash
+      mp <- except $ lookupMintingPolicy mpHash mpsMap >>=
```

Comment:
The code will be much simpler after the hash is replaced with the script itself.